### PR TITLE
Issue #331: add template backward compatibility

### PR DESCRIPTION
**Pull request by gaurabdg**
#331

### CODE REVIEW

**Comment by romani** [comment]
@gaurabdg, please share screenshots on how it looks like. Did you test backwards compatibility ?

CI is not happy.

**Comment by gaurabdg** [comment]
> @gaurabdg, please share screenshots on how it looks like.
Screenshot 2020-09-27 at 2 58 34 AM

> Did you test backwards compatibility ?

Yes, created config and ran `mvn sonar:sonar`.

> CI (sonar) is not happy.

@muhlba91 Please look into this, I just reintroduced old code, which are giving errors.

**Comment by muhlba91** [comment]
@gaurabdg we had similar problems already and copied over code from this library. please copy-paste the methods used into sonar-checkstyle in the same style like we did with https://github.com/checkstyle/sonar-checkstyle/blob/4.33/src/main/java/org/sonar/plugins/checkstyle/CheckstyleRulesDefinition.java#L73 (`CheckstyleRulesDefinition::loadNames`).

**Comment by gaurabdg** [comment]
@muhlba91 I dont think you are understanding the problem, the code(which existed in master before I removed) used is depreacted.
I am adding `SuppressWarning` annotation to pass sonar CI.

File: src/test/java/org/sonar/plugins/checkstyle/CheckstyleRulesDefinitionTest.java [hunk 1]
```diff
@@ -61,7 +61,7 @@ public void test() {
         assertThat(repository.language()).isEqualTo("java");
 
         final List<RulesDefinition.Rule> rules = repository.rules();
-        assertThat(rules).hasSize(179);
+        assertThat(rules).hasSize(353);
```
Comment (romani) [inline src/test/java/org/sonar/plugins/checkstyle/CheckstyleRulesDefinitionTest.java:64]: this should be 179 * 2 = 358 - (amount of Checks that are new)

Looks like we have 5 new Checks, but I see only 4 new checks https://checkstyle.org/releasenotes.html 

IllegalIdentifierName
RecordComponentNumberCheck
RecordTypeParameterNameCheck
PatternVariableName.

https://checkstyle.org/releasenotes.html#Release_8.33 we introduced NoCodeInFile and it is old metadata, so most likely has to be duplicated.

Please help to explain why `353`.

**Comment by muhlba91** [comment]
@gaurabdg yes, the code in sslr bridge is deprecated and they will remove it at some point. since we rely on that code part, we should move it to our code base and not rely on sslr bridge to provide it. once it‘s in our code base, we don‘t have to deprecate it.

problem will be: in a future release they will just delete the code. we still need it, so we should copy the code from sslr bridge to our code base.
that was the same with loadNames. they simply removed it (as far as i still remember) and we had to copy it. i‘d rather not like running into the same situation if i can avoid it already...

however, i also see this not as part of this particular issue to solve. if there‘s time for you, please copy it over, otherwise, i‘ll perform the transition in a future commit. :) 

good to go from my side once the check count question is resolved.

**Comment by romani** [comment]
Lets postpone deprecation nuance for later. We need to make it work first.

@muhlba91 , please confirm one more time that your existing sonar config is migrated correctly to new version of sonar.

**Comment by romani** [comment]
Travis CI faililng with:
```
[INFO] --- maven-checkstyle-plugin:3.1.1:check (checkstyle-check) @ checkstyle-sonar-plugin ---
[INFO] There is 1 error reported by Checkstyle 8.35 with https://raw.githubusercontent.com/checkstyle/checkstyle/checkstyle-8.35/config/checkstyle_checks.xml ruleset.
[ERROR] src/main/java/org/sonar/plugins/checkstyle/CheckstyleRulesDefinition.java:[60,23]
 (annotation) SuppressWarnings: The warning 'squid:S1874' cannot be suppressed at this location.
            Only few javac warnings are allowed to suppress.            
If try to suppress checkstyle/pmd/..... violation please do this in their config file.            
If you try to suppress IntelliJ IDEA inspection,            
please use javadoc block tag @noinspection
``` 

I am ok to suppress this violation for now in suppression.xml .

**Comment by romani** [comment]
@gaurabdg , please reply.

File: src/test/java/org/sonar/plugins/checkstyle/CheckstyleRulesDefinitionTest.java (hunk 1 above)
Comment (romani) [reply src/test/java/org/sonar/plugins/checkstyle/CheckstyleRulesDefinitionTest.java:64]: I found missed by me Check - https://checkstyle.org/config_javadoc.html#RequireEmptyLineBeforeBlockTagGroup
there was mistake in release notes, bad issue title.

fixed.

Comment (romani) [reply src/test/java/org/sonar/plugins/checkstyle/CheckstyleRulesDefinitionTest.java]: I updated code to validate counts more clearly

**Review by romani — APPROVED** [review]
@muhlba91 , please finalize review and testing.

File: src/main/java/org/sonar/plugins/checkstyle/metadata/CheckstyleMetadata.java [hunk 2]
```diff
@@ -100,9 +100,10 @@ public void createRulesWithMetadata() {
                     final SonarRulePropertyLoader.AdditionalRuleProperties additionalDetails =
                             additionalRuleData.get(check.getKey());
                     final RulesDefinition.NewRule rule =
-                            repository.createRule(moduleDetails.getFullQualifiedName());
+                            repository.createRule(moduleDetails.getFullQualifiedName()
+                                    + "template");
```
Comment (romani) [inline src/main/java/org/sonar/plugins/checkstyle/metadata/CheckstyleMetadata.java:104]: the only minor item it have for this line.
it make rule id looks like 
`com.puppycrawl.tools.checkstyle.checks.naming.AbbreviationAsWordInNameChecktemplate`
but probably better to do
`com.puppycrawl.tools.checkstyle.checks.naming.AbbreviationAsWordInNameCheckTemplate`

@muhlba91 , please share your thoughts on this.

**Comment by muhlba91** [comment]
@romani @gaurabdg i was not able to simply run sonar after updating:
```
Caused by: java.lang.IllegalArgumentException: given name empty
        at com.puppycrawl.tools.checkstyle.utils.TokenUtil.getTokenId(TokenUtil.java:174)
        at com.puppycrawl.tools.checkstyle.TreeWalker.registerCheck(TreeWalker.java:212)
        at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:130)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:201)
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:477)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:201)
        at org.sonar.plugins.checkstyle.CheckstyleExecutor.executeWithClassLoader(CheckstyleExecutor.java:102)
        ... 217 more
```

what i did is this: start SQ with 8.35 (so the previous version), bulk enable all checkstyle rules (no templates though), run analysis, update to 8.36-SNAPSHOT, restart SQ, run analysis.

i did not receive any further output from `gradle sonar --stacktrace` but i'll add the generated checkstyle.xml. @romani could you please have a quick look at it?

other than that, if we cannot figure out a good way we must decide on either a major version release to indicate incompatibility or mix and match forever but with limiting a user's functionality and also having to maintain those old checks manually. any suggestions on this one?
personally, i do not see a problem in a major version upgrade and breaking compatibility and i probably think it will be inevitable in the future.

[checkstyle.txt](https://github.com/checkstyle/sonar-checkstyle/files/5348270/checkstyle.txt)

**Comment by romani** [comment]
@gaurabdg, please help to figure out what is problem.

**Comment by rnveach** [comment]
The config with the issue is:
````
		<module name="DescendantToken">
			<property name="severity" value="info" />
			<property name="maximumDepth" value="2147483647" />
			<property name="maximumNumber" value="2147483647" />
			<property name="tokens" value="empty" />
			<property name="minimumNumber" value="0" />
			<property name="minimumDepth" value="0" />
			<property name="sumTokenCounts" value="false" />
		</module>
````

Notice the line `<property name="tokens" value="empty" />`.

I assume this relates back to https://github.com/checkstyle/checkstyle/blob/39b02e16d55a081ba779c0bdb2318ab30eaea515/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/DescendantTokenCheck.xml#L49

**Comment by romani** [comment]
@gaurabdg, please help to fix metadata file

**Comment by strkkk** [comment]
@romani I can suggest following. In DescendantTokenCheck value for tokens property will be set to empty string (or any value that should be treated as empty).
Then, add to setTokens method filtering for empty strings
https://github.com/checkstyle/checkstyle/blob/58bae0f3defcf7ba4a9fa69bed97aadcb35eb2ae/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java#L95-L97

By doing that, this line in registerCheck method
https://github.com/checkstyle/checkstyle/blob/58bae0f3defcf7ba4a9fa69bed97aadcb35eb2ae/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L201

will return empty set of tokens, which is correct.

It seems to me like a very small fix with minimal regression possible.

**Comment by romani** [comment]
I tested suggested by me fix in meta see PR: https://github.com/checkstyle/checkstyle/pull/8917

Screenshot from 2020-10-22 07-22-02
Screenshot from 2020-10-22 07-22-45

**Comment by romani** [comment]
@muhlba91 , please build checkstyle from my PR `mvn install -Pno-validations` and use it this branch.
Your extra help with testing will help a lot. If any problems found I can fix them before next checkstyle release (Oct 31).

**Comment by muhlba91** [comment]
fyi: i‘ll run my tests most likely tomorrow evening. unfortunately, those weeks are busier on my side.

**Comment by romani** [comment]
Thanks a lot.
I will release checkstyle library with required for this PR update update on this weekend.

**Comment by muhlba91** [comment]
tested and looking good from my side:
- 8.35: enable all rules & run analysis
- update CS plugin & run analysis
- enable all rules of snapshot & run analysis
- enable some template rules & run analysis

**Comment by romani** [comment]
@muhlba91 , I released https://checkstyle.org/releasenotes.html#Release_8.37 , looks like we need to merge this PR
and upgrade to 8.37 and do final test.

**Pull request merged**